File: python/caskada/_observation.py
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
# Copyright (c) 2025, Victor Duarte
# Run accounting, event publication, and observable failure fences.
from __future__ import annotations

import inspect
import time
from collections import deque
from collections.abc import Sequence
from typing import (
    Any,
    Literal,
    TypeAlias,
    cast,
)

from ._contracts import (
    MAX_SAFE_INTEGER,
    AttemptFenceTarget,
    CallbackDisposition,
    CallbackFinishedEvent,
    CallbackFinishedPayload,
    CallbackStartedEvent,
    CallbackStartedPayload,
    CancellationFencedEvent,
    CancellationFencedPayload,
    Failure,
    FailureFencedEvent,
    FailureFencedPayload,
    FailureRecordedEvent,
    FailureRecordedPayload,
    Observer,
    ObserverDiagnostic,
    Phase,
    RunEvent,
    RunFenceTarget,
    RunFinishedEvent,
    RunFinishedPayload,
    RunStats,
    ScopeFenceTarget,
    ScopeFinishedEvent,
    ScopeFinishedPayload,
    ScopeStartedEvent,
    ScopeStartedPayload,
)
from ._failures import _FailureFence, _ProducedFailure
from ._timing import _CancellationSource, _dispose_invalid_sync_result

_EventSpec: TypeAlias = tuple[type[Any], object]


class _EventPublisher:
    __slots__ = (
        "_diagnostics",
        "_disabled",
        "_observer",
        "_pending",
        "_publishing",
        "_run_cancellation_published",
        "_run_id",
        "_sequence",
        "_terminal",
    )

    def __init__(self, run_id: str, observer: Observer | None) -> None:
        self._run_id = run_id
        self._observer = observer
        self._sequence = 0
        self._diagnostics: list[ObserverDiagnostic] = []
        self._disabled = False
        self._publishing = False
        self._pending: deque[tuple[_EventSpec, ...]] = deque()
        self._run_cancellation_published = False
        self._terminal = False

    @property
    def diagnostics(self) -> tuple[ObserverDiagnostic, ...]:
        return tuple(self._diagnostics)

    @property
    def publishing(self) -> bool:
        return self._publishing

    def publish(self, event_type: type[Any], payload: object) -> None:
        self.publish_bundle(((event_type, payload),))

    def publish_bundle(self, specs: Sequence[_EventSpec]) -> None:
        if not specs:
            return
        captured = tuple(specs)
        if self._publishing:
            self._pending.append(captured)
            return
        self._pending.append(captured)
        self._publishing = True
        try:
            while self._pending:
                bundle = self._pending.popleft()
                events: list[RunEvent] = []
                for event_type, payload in bundle:
                    self._sequence += 1
                    events.append(event_type(self._sequence, self._run_id, payload))
                for event in events:
                    self._invoke(event)
        finally:
            self._publishing = False

    def reject_reentrant_report(self) -> None:
        if not self._publishing or self._disabled:
            return
        self._disable(
            self._sequence,
            "Observer reentrancy disabled",
            None,
        )

    def publish_run_cancellation(self, reason: object, deadline: bool) -> None:
        if self._run_cancellation_published:
            return
        self._run_cancellation_published = True
        self.publish(
            CancellationFencedEvent,
            CancellationFencedPayload(RunFenceTarget(), reason, deadline),
        )

    def mark_run_cancellation_published(self) -> None:
        self._run_cancellation_published = True

    @property
    def terminal(self) -> bool:
        return self._terminal

    def mark_terminal(self) -> None:
        self._terminal = True

    def _invoke(self, event: RunEvent) -> None:
        observer = self._observer
        if observer is None or self._disabled:
            return
        try:
            result = observer(event)
        except BaseException as cause:
            self._disable(event.sequence, "Observer raised", cause)
            return
        try:
            asynchronous = inspect.isawaitable(result) or inspect.isasyncgen(result)
        except BaseException as cause:
            self._disable(event.sequence, "Observer result inspection failed", cause)
            return
        if asynchronous:
            _dispose_invalid_sync_result(result)
            self._disable(
                event.sequence,
                "Observer must return synchronously",
                None,
            )

    def _disable(
        self,
        sequence: int,
        message: str,
        cause: BaseException | None,
    ) -> None:
        if self._disabled:
            return
        self._disabled = True
        self._diagnostics.append(ObserverDiagnostic(sequence, message, cause))


class _RunAccounting:
    """Owns run counters, portable identities, and the frozen stats snapshot."""

    def __init__(self, started_ns: int) -> None:
        self.started_ns = started_ns
        self.terminal_ns: int | None = None
        self.next_activation_id = 2
        self.next_scope_id = 2
        self.next_terminal_sequence = 1
        self.activations = 1
        self.attempts = 0
        self.transitions = 0
        self.retries = 0
        self.reports = 0
        self.scopes = 0
        self.ready = 0
        self.peak_ready = 0

    def allocate_activation_id(self) -> int:
        value = self.next_activation_id
        self.next_activation_id += 1
        self.activations += 1
        return value

    def allocate_terminal_sequence(self) -> int:
        value = self.next_terminal_sequence
        self.next_terminal_sequence += 1
        return value

    def stats(self, peak_callbacks: int) -> RunStats:
        if self.terminal_ns is None:
            self.terminal_ns = time.monotonic_ns()
        duration_ms = min(
            MAX_SAFE_INTEGER,
            (self.terminal_ns - self.started_ns) // 1_000_000,
        )
        return RunStats(
            activations=self.activations,
            attempts=self.attempts,
            transitions=self.transitions,
            retries=self.retries,
            reports=self.reports,
            scopes=self.scopes,
            peak_ready=self.peak_ready,
            peak_callbacks=peak_callbacks,
            duration_ms=duration_ms,
        )


class _RunObserver:
    """Owns event ordering, fences, and observer-visible run state."""

    def __init__(
        self,
        publisher: _EventPublisher,
        cancellation: _CancellationSource,
        runtime_scopes: dict[int, Any],
    ) -> None:
        self.publisher = publisher
        self.cancellation = cancellation
        self.runtime_scopes = runtime_scopes
        self.failure_fence: _ProducedFailure | None = None
        self.recorded_failures: set[int] = set()
        self.run_fence_published = False
        self.cancellation_fence_published = False
        self.scope_fences_published: set[int] = set()
        self.attempt_fences_published: set[tuple[int, int, int]] = set()

    def _scope_started_spec(self, scope: Any) -> _EventSpec:
        return (
            ScopeStartedEvent,
            ScopeStartedPayload(
                scope_id=scope.scope_id,
                parent_scope_id=None if scope.parent is None else scope.parent.scope_id,
                owner_activation_id=scope.owner_activation_id,
                entry_activation_id=scope.entry_activation_id,
                entry_element_id=scope.definition.entry_element_id,
                flow_element_id=scope.owner_placement.element_id,
                depth=scope.depth,
            ),
        )

    def _scope_finished_spec(
        self,
        scope: Any,
        status: Literal["completed", "failed", "cancelled", "abandoned"],
    ) -> _EventSpec:
        return (
            ScopeFinishedEvent,
            ScopeFinishedPayload(
                scope_id=scope.scope_id,
                status=status,
                terminal_sequences=(
                    scope.finished_terminal_sequences
                    if scope.finished_terminal_sequences is not None
                    else tuple(terminal.sequence for terminal in scope.terminals)
                ),
            ),
        )

    def _mark_scope_finished(
        self,
        scope: Any,
        status: Literal["completed", "failed", "cancelled", "abandoned"],
    ) -> _EventSpec | None:
        if scope.finished:
            return None
        scope.finished = True
        return self._scope_finished_spec(scope, status)

    @staticmethod
    def _capture_scope_finish_terminals(scope: Any) -> None:
        if scope.finished_terminal_sequences is None:
            scope.finished_terminal_sequences = tuple(
                terminal.sequence for terminal in scope.terminals
            )

    def _publish_terminal(
        self,
        root: Any,
        status: Literal["completed", "failed", "cancelled", "abandoned"],
    ) -> None:
        self.publisher.mark_terminal()
        scope_status: Literal["completed", "failed", "cancelled", "abandoned"] = status
        specs: list[_EventSpec] = []
        for scope in sorted(
            self.runtime_scopes.values(),
            key=lambda candidate: (-candidate.depth, candidate.scope_id),
        ):
            spec = self._mark_scope_finished(scope, scope_status)
            if spec is not None:
                specs.append(spec)
        specs.append((RunFinishedEvent, RunFinishedPayload(status)))
        self.publisher.publish_bundle(specs)

    def _publish_callback_started(
        self,
        scope: Any,
        activation_id: int,
        parent_activation_id: int | None,
        element_id: int,
        phase: Phase,
        attempt: int | None,
    ) -> None:
        self.publisher.publish(
            CallbackStartedEvent,
            CallbackStartedPayload(
                scope.scope_id,
                activation_id,
                parent_activation_id,
                element_id,
                phase,
                attempt,
            ),
        )

    def _failure_record_spec(self, failure: Failure) -> _EventSpec | None:
        if failure.failure_id in self.recorded_failures:
            return None
        self.recorded_failures.add(failure.failure_id)
        return FailureRecordedEvent, FailureRecordedPayload(failure)

    def _publish_callback_finished(
        self,
        *,
        scope_id: int,
        activation_id: int,
        phase: Phase,
        attempt: int | None,
        disposition: CallbackDisposition,
        failures: Sequence[Failure] = (),
    ) -> None:
        specs = [
            spec
            for failure in failures
            if (spec := self._failure_record_spec(failure)) is not None
        ]
        specs.append(
            (
                CallbackFinishedEvent,
                CallbackFinishedPayload(
                    scope_id,
                    activation_id,
                    phase,
                    attempt,
                    disposition,
                ),
            )
        )
        self.publisher.publish_bundle(specs)

    def _publish_failure_recorded(self, failure: Failure) -> None:
        spec = self._failure_record_spec(failure)
        if spec is not None:
            self.publisher.publish_bundle((spec,))

    def _publish_scope_failure_fence(
        self,
        scope: Any,
        failure: Failure,
    ) -> None:
        if scope.scope_id in self.scope_fences_published:
            return
        self.scope_fences_published.add(scope.scope_id)
        specs: list[_EventSpec] = []
        record = self._failure_record_spec(failure)
        if record is not None:
            specs.append(record)
        specs.extend(
            (
                (
                    FailureFencedEvent,
                    FailureFencedPayload(ScopeFenceTarget(scope.scope_id), failure),
                ),
                (
                    CancellationFencedEvent,
                    CancellationFencedPayload(
                        ScopeFenceTarget(scope.scope_id),
                        "scope_failed",
                        False,
                    ),
                ),
            )
        )
        self.publisher.publish_bundle(specs)

    def _publish_run_failure_fence(self, failure: Failure) -> None:
        if self.run_fence_published:
            return
        self.run_fence_published = True
        self.cancellation_fence_published = True
        self.failure_fence = self.failure_fence or _ProducedFailure(failure)
        self.cancellation.cancel(_FailureFence(self.failure_fence))
        self.publisher.mark_run_cancellation_published()
        specs: list[_EventSpec] = []
        record = self._failure_record_spec(failure)
        if record is not None:
            specs.append(record)
        specs.extend(
            (
                (
                    FailureFencedEvent,
                    FailureFencedPayload(RunFenceTarget(), failure),
                ),
                (
                    CancellationFencedEvent,
                    CancellationFencedPayload(
                        RunFenceTarget(),
                        "run_failed",
                        False,
                    ),
                ),
            )
        )
        self.publisher.publish_bundle(specs)

    def _publish_run_cancellation_if_needed(self) -> None:
        if not self.cancellation.cancelled or self.cancellation_fence_published:
            return
        self.cancellation_fence_published = True
        self.publisher.publish_run_cancellation(
            self.cancellation.reason,
            self.cancellation.deadline,
        )

    def _publish_attempt_timeout(
        self,
        context: Any,
        failure: Failure,
    ) -> None:
        target_key = (
            context._scope_id,
            context._activation_id,
            cast(int, context._attempt),
        )
        if target_key in self.attempt_fences_published:
            return
        self.attempt_fences_published.add(target_key)
        spec = self._failure_record_spec(failure)
        specs: list[_EventSpec] = [] if spec is None else [spec]
        specs.append(
            (
                CancellationFencedEvent,
                CancellationFencedPayload(
                    AttemptFenceTarget(
                        context._scope_id,
                        context._activation_id,
                        target_key[2],
                    ),
                    "attempt_timeout",
                    False,
                ),
            )
        )
        self.publisher.publish_bundle(specs)

File: typescript/internal/observation.ts
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// Copyright (c) 2025, Victor Duarte

// Run accounting, event publication, and observable failure fences.

import { MAX_SAFE_INTEGER } from './contracts.js'
import { ProducedFailure } from './failures.js'
import { disposeNativePromise, RuntimeCancellation } from './timing.js'

import type { Failure, Observer, ObserverDiagnostic, Phase, RunEvent, RunStats, Terminal } from './contracts.js'

export type EventSpec = {
  readonly kind: RunEvent['kind']
  readonly payload: unknown
}

export class RunAccounting {
  nextActivationId = 2
  nextScopeId = 2
  nextTerminalSequence = 1
  activations = 1
  attempts = 0
  transitions = 0
  retries = 0
  reports = 0
  scopes = 0
  ready = 0
  peakReady = 0
  private terminalMs: number | undefined

  constructor(private readonly startedMs: number) {}

  allocateActivationId(): number {
    const value = this.nextActivationId
    this.nextActivationId += 1
    this.activations += 1
    return value
  }

  allocateTerminalSequence(): number {
    const value = this.nextTerminalSequence
    this.nextTerminalSequence += 1
    return value
  }

  stats(peakCallbacks: number): RunStats {
    this.terminalMs ??= performance.now()
    const durationMs = Math.min(MAX_SAFE_INTEGER, Math.max(0, Math.floor(this.terminalMs - this.startedMs)))
    return Object.freeze({
      activations: this.activations,
      attempts: this.attempts,
      transitions: this.transitions,
      retries: this.retries,
      reports: this.reports,
      scopes: this.scopes,
      peakReady: this.peakReady,
      peakCallbacks,
      durationMs,
    })
  }
}

export class EventPublisher {
  private sequence = 0
  private readonly observerDiagnostics: ObserverDiagnostic[] = []
  private disabled = false
  private publishing = false
  private runCancellationPublished = false
  private terminal = false
  private readonly pending: Array<readonly EventSpec[]> = []

  constructor(
    private readonly runId: string,
    private readonly observer: Observer | undefined,
  ) {}

  get diagnostics(): readonly ObserverDiagnostic[] {
    return Object.freeze(this.observerDiagnostics.slice())
  }

  get isPublishing(): boolean {
    return this.publishing
  }

  publish(kind: RunEvent['kind'], payload: unknown): void {
    this.publishBundle([{ kind, payload }])
  }

  publishBundle(specs: readonly EventSpec[]): void {
    if (specs.length === 0) return
    this.pending.push(specs.slice())
    if (this.publishing) return
    this.publishing = true
    try {
      while (this.pending.length > 0) {
        const bundle = this.pending.shift()!
        const events = bundle.map((spec): RunEvent => {
          this.sequence += 1
          return Object.freeze({ sequence: this.sequence, runId: this.runId, kind: spec.kind, payload: spec.payload }) as RunEvent
        })
        for (const event of events) this.invoke(event)
      }
    } finally {
      this.publishing = false
    }
  }

  rejectReentrantReport(): void {
    if (!this.publishing || this.disabled) return
    this.disable(this.sequence, 'Observer reentrancy disabled', null)
  }

  publishRunCancellation(reason: unknown, deadline: boolean): void {
    if (this.runCancellationPublished) return
    this.runCancellationPublished = true
    this.publish('cancellation_fenced', Object.freeze({ target: Object.freeze({ kind: 'run' }), reason, deadline }))
  }

  markRunCancellationPublished(): void {
    this.runCancellationPublished = true
  }

  get terminalCommitted(): boolean {
    return this.terminal
  }

  markTerminal(): void {
    this.terminal = true
  }

  private invoke(event: RunEvent): void {
    if (this.observer === undefined || this.disabled) return
    let result: unknown
    try {
      result = this.observer(event)
    } catch (cause) {
      this.disable(event.sequence, 'Observer raised', cause)
      return
    }
    if ((typeof result !== 'object' || result === null) && typeof result !== 'function') return
    let then: unknown
    try {
      then = Reflect.get(result, 'then')
    } catch (cause) {
      this.disable(event.sequence, 'Observer result inspection failed', cause)
      return
    }
    if (typeof then !== 'function') return
    disposeNativePromise(result)
    this.disable(event.sequence, 'Observer must return synchronously', null)
  }

  private disable(eventSequence: number, message: string, cause: unknown | null): void {
    if (this.disabled) return
    this.disabled = true
    this.observerDiagnostics.push(Object.freeze({ eventSequence, message, cause }))
  }
}

type ObservedScope = {
  readonly scopeId: number
  readonly parent: { readonly scopeId: number } | null
  readonly ownerActivationId: number
  readonly entryActivationId: number
  readonly definition: { readonly entryElementId: number }
  readonly ownerPlacement: { readonly elementId: number }
  readonly depth: number
  readonly terminals: readonly Terminal[]
  finishedTerminalSequences?: readonly number[] | undefined
  finished: boolean
}

type ObservedAttempt = {
  readonly scopeId: number
  readonly activationId: number
  readonly attempt: number | null
}

export class RunObserver {
  failureFence: ProducedFailure | undefined
  private readonly recordedFailures = new Set<number>()
  private readonly scopeFencesPublished = new Set<number>()
  private readonly attemptFencesPublished = new Set<string>()
  private runFencePublished = false
  private cancellationFencePublished = false

  constructor(
    readonly publisher: EventPublisher,
    private readonly cancellation: RuntimeCancellation,
    private readonly runtimeScopes: ReadonlyMap<number, ObservedScope>,
  ) {}

  scopeStartedSpec(scope: ObservedScope): EventSpec {
    return {
      kind: 'scope_started',
      payload: Object.freeze({
        scopeId: scope.scopeId,
        parentScopeId: scope.parent?.scopeId ?? null,
        ownerActivationId: scope.ownerActivationId,
        entryActivationId: scope.entryActivationId,
        entryElementId: scope.definition.entryElementId,
        flowElementId: scope.ownerPlacement.elementId,
        depth: scope.depth,
      }),
    }
  }

  scopeFinishedSpec(scope: ObservedScope, status: 'completed' | 'failed' | 'cancelled' | 'abandoned'): EventSpec {
    return {
      kind: 'scope_finished',
      payload: Object.freeze({
        scopeId: scope.scopeId,
        status,
        terminalSequences: scope.finishedTerminalSequences ?? Object.freeze(scope.terminals.map((terminal) => terminal.sequence)),
      }),
    }
  }

  markScopeFinished(scope: ObservedScope, status: 'completed' | 'failed' | 'cancelled' | 'abandoned'): EventSpec | undefined {
    if (scope.finished) return undefined
    scope.finished = true
    return this.scopeFinishedSpec(scope, status)
  }

  captureScopeFinishTerminals(scope: ObservedScope): void {
    scope.finishedTerminalSequences ??= Object.freeze(scope.terminals.map((terminal) => terminal.sequence))
  }

  publishTerminal(status: 'completed' | 'failed' | 'cancelled' | 'abandoned'): void {
    this.publisher.markTerminal()
    const scopes = [...this.runtimeScopes.values()].sort((left, right) => right.depth - left.depth || left.scopeId - right.scopeId)
    const specs: EventSpec[] = []
    for (const scope of scopes) {
      const spec = this.markScopeFinished(scope, status)
      if (spec !== undefined) specs.push(spec)
    }
    specs.push({ kind: 'run_finished', payload: Object.freeze({ status }) })
    this.publisher.publishBundle(specs)
  }

  publishCallbackStarted(
    scope: ObservedScope,
    activationId: number,
    parentActivationId: number | null,
    elementId: number,
    phase: Phase,
    attempt: number | null,
  ): void {
    this.publisher.publish(
      'callback_started',
      Object.freeze({ scopeId: scope.scopeId, activationId, parentActivationId, elementId, phase, attempt }),
    )
  }

  failureRecordSpec(failure: Failure): EventSpec | undefined {
    if (this.recordedFailures.has(failure.failureId)) return undefined
    this.recordedFailures.add(failure.failureId)
    return { kind: 'failure_recorded', payload: Object.freeze({ failure }) }
  }

  publishCallbackFinished(
    scopeId: number,
    activationId: number,
    phase: Phase,
    attempt: number | null,
    disposition:
      | { readonly kind: 'outcome'; readonly outcome: 'route' | 'fanout' | 'end' | 'forward' | 'unhandled' }
      | { readonly kind: 'failure'; readonly failure: Failure }
      | { readonly kind: 'discarded' },
    failures: readonly Failure[] = [],
  ): void {
    const specs: EventSpec[] = []
    for (const failure of failures) {
      const spec = this.failureRecordSpec(failure)
      if (spec !== undefined) specs.push(spec)
    }
    specs.push({
      kind: 'callback_finished',
      payload: Object.freeze({ scopeId, activationId, phase, attempt, disposition: Object.freeze(disposition) }),
    })
    this.publisher.publishBundle(specs)
  }

  publishFailureRecorded(failure: Failure): void {
    const spec = this.failureRecordSpec(failure)
    if (spec !== undefined) this.publisher.publishBundle([spec])
  }

  publishScopeFailureFence(scope: ObservedScope, failure: Failure): void {
    if (this.scopeFencesPublished.has(scope.scopeId)) return
    this.scopeFencesPublished.add(scope.scopeId)
    const specs: EventSpec[] = []
    const record = this.failureRecordSpec(failure)
    if (record !== undefined) specs.push(record)
    specs.push(
      {
        kind: 'failure_fenced',
        payload: Object.freeze({ target: Object.freeze({ kind: 'scope', scopeId: scope.scopeId }), failure }),
      },
      {
        kind: 'cancellation_fenced',
        payload: Object.freeze({
          target: Object.freeze({ kind: 'scope', scopeId: scope.scopeId }),
          reason: 'scope_failed',
          deadline: false,
        }),
      },
    )
    this.publisher.publishBundle(specs)
  }

  publishRunFailureFence(failure: Failure): void {
    if (this.runFencePublished) return
    this.runFencePublished = true
    this.cancellationFencePublished = true
    this.failureFence ??= new ProducedFailure(failure)
    this.cancellation.cancel(Object.freeze({ type: 'failure_fence' as const, produced: this.failureFence }))
    this.publisher.markRunCancellationPublished()
    const specs: EventSpec[] = []
    const record = this.failureRecordSpec(failure)
    if (record !== undefined) specs.push(record)
    specs.push(
      { kind: 'failure_fenced', payload: Object.freeze({ target: Object.freeze({ kind: 'run' }), failure }) },
      {
        kind: 'cancellation_fenced',
        payload: Object.freeze({ target: Object.freeze({ kind: 'run' }), reason: 'run_failed', deadline: false }),
      },
    )
    this.publisher.publishBundle(specs)
  }

  publishRunCancellationIfNeeded(): void {
    if (!this.cancellation.cancelled || this.cancellationFencePublished) return
    this.cancellationFencePublished = true
    this.publisher.publishRunCancellation(this.cancellation.reason, this.cancellation.deadline)
  }

  publishAttemptTimeout(context: ObservedAttempt, failure: Failure): void {
    const targetKey = `${context.scopeId}:${context.activationId}:${context.attempt!}`
    if (this.attemptFencesPublished.has(targetKey)) return
    this.attemptFencesPublished.add(targetKey)
    const specs: EventSpec[] = []
    const record = this.failureRecordSpec(failure)
    if (record !== undefined) specs.push(record)
    specs.push({
      kind: 'cancellation_fenced',
      payload: Object.freeze({
        target: Object.freeze({
          kind: 'attempt',
          scopeId: context.scopeId,
          activationId: context.activationId,
          attempt: context.attempt!,
        }),
        reason: 'attempt_timeout',
        deadline: false,
      }),
    })
    this.publisher.publishBundle(specs)
  }
}
